approov-service/src/main/java/io/approov/service/okhttp/ApproovService.java

```diff
@@ -862,6 +903,69 @@ else if (approovResults.getStatus() != Approov.TokenFetchStatus.UNKNOWN_KEY)
+                return chain.proceed(request);
+            }
+            Log.d(TAG, "Signing message with configuration for domain: "  + domain + ":  " + (messageSigningConfig.getSignApproovToken() ? ", " + approovTokenHeader +  " **** " : "") + " headers " + messageSigningConfig.getSignedHeaders() +
+                    (messageSigningConfig.getSignBody() ? ", body" : ""));
+            // build the message to sign, consisting of the URL, the names and values of the included headers and
+            // the body, if enabled, where each entry is separated from the next by a newline character
+            StringBuilder message = new StringBuilder();
+            //  1. add the URL to the message, followed by a newline
+            message.append(request.url());
+            message.append("\n");
+            // 2. Add the Method to the message
+            message.append(request.method());
+            message.append("\n");
+            // 3. add the Approov token to the message, if required (TODO: Note we add the approov token and the prefix!?????)
+            if (messageSigningConfig.signApproovToken) {
+                if (approovResults.getToken() != null) {
+                    message.append(approovTokenPrefix + approovResults.getToken());
+                }
+                message.append("\n");
+            }
+            // 4. add the required headers to the message as 'headername:headervalue', where the headername is in
+            // lowercase
+            if (messageSigningConfig.getSignedHeaders() != null) {
+                for (String header : messageSigningConfig.signedHeaders) {
+                    // add one headername:headervalue\n entry for each header value to be included in the signature
+                    List<String> values = request.headers(header);
+                    for (String value : values) {
+                        message.append(header.toLowerCase()).append(":");
+                        if (value != null) {
+                            message.append(value);
+                        }
+                        message.append("\n");
+                    }
+                }
+            }
+            // add the body to the message
+            if (messageSigningConfig.getSignBody()) {
+                okhttp3.RequestBody body = request.body();
+                if (body != null) {
+                    Buffer buffer = new Buffer();
+                    body.writeTo(buffer);
+                    message.append(buffer.readUtf8());
+                }
+            }
+            // compute the signature and add it to the request (passing on any exceptions that may occur)
+            if (messageSigningConfig.targetHeader != null && !messageSigningConfig.targetHeader.isEmpty()) {
+                String signature = ApproovService.getMessageSignature(message.toString());
+                request = request.newBuilder().header(messageSigningConfig.targetHeader, signature).build();
```

Comment:
Once we have the signature, call the 'generateHeaderValue' method on the MessageSigningConfig to generate the string value for the header instead of just adding the signature directly.